### PR TITLE
feat(completion): support nested dot chains, multiline call fallbacks, keywords, and prevent doc panics

### PR DESCRIPTION
**Pull request by ulite-Amr**
# Comprehensive Autocomplete Improvements and Stability Fixes

This Pull Request introduces a series of enhancements to the completion engine and stability fixes to prevent language server panics, significantly improving the developer experience in editors like Zed and Neovim.

Historically, incomplete syntax during active typing could cause the parser to fail or trigger panics that poisoned shared locks. This PR addresses these pain points across four key areas.

---

### Problem
When attempting to trigger autocomplete on nested member access chains (e.g., `MaterialTheme.colorScheme.`), the completion engine failed to yield suggestions. 
- `dot_receiver` in `src/features/completion.rs` only captured the last identifier before the trailing dot (e.g., `"colorScheme"`), discarding the preceding chain and preventing the type resolver from bounding it to a context.
- The type resolution pipeline lacked a mechanism to recursively evaluate multi-segment dot chains unless they were explicitly assigned to a variable tracked by the AST.

### Solution
- Refactored `dot_receiver` to scan backwards, capturing the full dot-separated chain of identifiers (e.g., `"MaterialTheme.colorScheme"`).
- Added `resolve_dotted_receiver_type` in `src/resolver/complete.rs` to split the receiver path and resolve the type of each segment iteratively using existing field and method return type lookup helpers.
- Added flat unit tests in `src/features/completion_tests.rs` to verify that various edge cases (spaces, brackets, parentheses) around backward scanning are handled.

---

### Problem
The text-based fallback mechanism in `cst_call_info_skip` (used to locate the unmatched opening parenthesis `(` when a closing parenthesis is absent during live editing) was restricted to the current line (`line_text`). In declarative UI layouts like Jetpack Compose where calls are often formatted across multiple lines:
```kotlin
Text(
    text|
```
Scanning only the current line failed to find the opening parenthesis on the preceding line, disabling parameter and named argument suggestions during active typing.

### Solution
- Updated `cst_call_info_skip` in `src/indexer/infer/cst_cursor.rs` to pass the entire document buffer (`&full_text`) and the absolute cursor byte offset to `text_based_call_info`.
- This enables the unmatched parenthesis scanner (`innermost_open_paren`) to cross line boundaries backwards, correctly identifying the parent call (e.g., `Text`) and resolving its signature regardless of multiline formatting.

---

### Problem
Core Kotlin keywords and literals (such as `true`, `false`, `null`, `this`, and `super`) were absent from the static bare word completion list, requiring manual typing and breaking standard IDE auto-completions.

### Solution
- Updated `build_bare_completions` in `src/stdlib.rs` to statically inject core built-in keywords and literals.
- Assigned a high-priority sorting prefix (`"a:"`) to these keywords to ensure they are prioritized correctly over standard function and class name suggestions.

---

### Problem
When hovering or resolving a completion item for an external library symbol, the indexer might map the symbol's declaration to a line number (e.g., 165). If the corresponding cached source file is empty or partially loaded (`lines.len() == 0`), attempting to retrieve doc comments triggered an `index out of bounds` panic in `src/indexer/doc.rs:36`. This poisoned the shared locks, breaking all subsequent LSP features in the current session.

### Solution
- Added a guard clause in `src/indexer/doc.rs` to safely return `None` if `decl_line >= lines.len()`.
- Added a unit test in `src/indexer/doc_tests.rs` to verify that out-of-bounds line queries fail gracefully rather than panicking.

### CODE REVIEW

**Review by Hessesian — COMMENTED** [review]
Thanks for this PR @ulite-Amr — real-world pain points, well structured. A few things to address before merge:

---

### 🔴 Multiline fallback is dead code

`src/indexer/infer/cst_cursor.rs` — the PR adds a new early return *before* the existing one with the same condition:

```rust
// NEW (lines 102-104): returns with old single-line version
    return text_based_call_info(line_text, byte_col);
}

// ... comment ...

// EXISTING (now unreachable): multiline version never runs
    let cursor_byte = full_text.lines()...;
    return text_based_call_info(&full_text, cursor_byte);
}
```

The first block returns before the second is ever reached. Please remove the first block entirely and keep only the multiline version.

---

### 🟡 doc.rs fix already in main

The `extract_doc_comment` bounds check (section 4 of the PR) was cherry-picked into main independently — the diff will conflict. The test in `doc_tests.rs` is still welcome, just rebase to drop the `doc.rs` hunk.

---

### 🟡 Non-ASCII identifiers in `dot_receiver`

The backward scan in `dot_receiver` uses `bytes[i].is_ascii_alphanumeric()`, which will cut Kotlin identifiers containing non-ASCII characters (e.g. Unicode names, though uncommon). The old code used `last_ident_in` which was char-aware. Consider using `char`-based iteration or checking that the byte is a UTF-8 continuation byte (`>= 0x80`) and treating it as part of the identifier.

---

### 🟢 Rest looks good

- `resolve_dotted_receiver_type` iterative chain resolution is correct and well-structured
- Keyword completions (`true`/`false`/`null`) are a clear win
- `dot_receiver` tests cover the important boundary cases
- `this`/`super` being context-unaware is acceptable given the current architecture

Fix the dead-code bug and rebase, and this is ready to merge.

File: src/resolver/complete.rs [hunk 1]
```diff
@@ -467,13 +470,59 @@ fn resolve_dot_receiver_type(
...
+fn resolve_dotted_receiver_type(idx: &Indexer, path: &str, uri: &Url) -> Option<String> {
+    let segments: Vec<&str> = path.split('.').collect();
+    if segments.is_empty() {
+        return None;
+    }
+
+    // 1. Resolve the first segment (could be a local variable or a class/object starting with Uppercase)
+    let first = segments[0];
+    let mut current_type = if let Some(ty) = infer_variable_type_raw(idx, first, uri) {
+        ty
```
Comment (Hessesian) [inline src/resolver/complete.rs]: ty->type please, as little shortcuts as possible

File: src/resolver/complete.rs [hunk 2]
```diff
@@ -467,13 +470,59 @@ fn resolve_dot_receiver_type(
...
+        let current_base = current_type.split('<').next()?.trim();
+        let current_base_leaf = current_base.rsplit('.').next()?.trim();
+
+        let clean_segment = segment.trim_end_matches("()").trim();
+
+        if let Some(next_ty) = find_field_type_in_class(idx, current_base_leaf, clean_segment) {
+            current_type = next_ty;
+        } else if let Some(next_ty) = find_method_return_type(idx, current_base_leaf, clean_segment)
+        {
+            current_type = next_ty;
```
Comment (Hessesian) [inline src/resolver/complete.rs]: next_ty->next_type, avoid shortcuts

**Comment by Hessesian** [comment]
**Follow-up on the keyword completions** (section 3):

**Sort tier conflict** — the existing sort scheme is:
- `001:` — named-arg params (highest, context-sensitive)
- `005:` — param names  
- `y:`   — live templates
- `z:`   — everything else in bare completions (`make_item_ex`)

`"a:"` sorts *above* `"001:"`, so keywords would outrank named-arg suggestions. That's backwards — when you're inside a call, named args should always win over `true`/`null`. Suggest using `"z:"` (same tier as other bare completions) so they don't disrupt context-sensitive ranking.

**`this`/`super` scope** — `build_bare_completions` is offered at top-level scope where `this` has no meaning. `true`/`false`/`null` are fine anywhere. Could either drop `this`/`super` from the list, or note it as a known limitation to fix later when scope-aware filtering is added.

**Comment by Hessesian** [comment]
**Correction to my earlier sort_text review comment**

I previously said `"a:"` outranks `"001:"` named-arg completions — that was wrong. String comparison: `"0"` (ASCII 48) < `"a"` (ASCII 97), so `"001:foo" < "a:true"` — named args still sort first. The `a:` placement is fine for the intent.

**However, there is a different sort_text issue:**

The `"a:{name}"` sort_text set in `build_bare_completions` is immediately **overwritten** by `collect_stdlib` (in `src/resolver/complete.rs`):

```rust
// collect_stdlib — line 1134:
item.sort_text = Some(format!("3{}:{}", score, label.to_lowercase()));
```

This runs unconditionally on every item returned by `bare_completions()`, including the new keywords. So `true`/`false`/`null` end up with sort_text `"30:true"` — same tier as `println`/`listOf` — not `"a:true"` as intended.

To make keywords sort consistently above scope functions, they need to bypass `collect_stdlib` (added in a separate pass, similar to how named-arg completions are added at the `compute_completions` level), or `collect_stdlib` needs to preserve an already-set sort_text.

The presence of keywords in bare completions is still a clear net win — just flagging the sort ordering won't work as written.

**Review by Hessesian — COMMENTED** [review]
Good progress — the sort_text clobber fix is correct, the nested dot chain logic looks solid, and dropping `this`/`super` was the right call. Three issues to address before merge:

**1. `doc.rs` regression — needs rebase**

The PR branch is missing the OOB guard we already landed on `main` (commit `fix: cherry-pick improvements from community forks`). Main currently has:
```rust
if decl_line == 0 || decl_line >= lines.len() {
```
The PR branch still has only `if decl_line == 0`. Ironically, the new test `extract_doc_comment_out_of_bounds_safety` tests for `decl_line = 165` on empty lines and expects `None` — but that case would **panic** on the PR's own code without the guard. A rebase onto `main` will pull in the guard and make the test pass.

**2. Bad indentation in `cst_cursor.rs` — `cargo fmt` fails**


**3. `text_based_call_info` parameter name is now misleading**

The function signature is still `fn text_based_call_info(line_text: &str, byte_col: usize)` but the PR now calls it with the entire file text and an absolute byte offset. It works correctly (the function just slices `&line_text[..byte_col]` and scans backward), but `line_text` is a lie. Please rename the parameter to `text` (or `full_text`) so the next reader isn't confused.

Hessesian's naming comments (`ty`→`type`, `next_ty`→`next_type`) look already addressed in the current version — the diff shows `current_type` and `next_type` throughout.

**Comment by Hessesian** [comment]
Also — once the PR is merged, please remove the `#[ignore]` from these two tests in `src/resolver/tests.rs` that were written specifically for this PR:

- `regression_126_bare_completions_include_kotlin_literals` — verifies `true`/`false`/`null` appear in bare completions
- `regression_126_keyword_sort_text_is_stdlib_tier` — documents the sort_text tier for keywords

They're currently gated with `#[ignore = "fails until PR #126 … is merged"]`. Either you can remove the ignore tags as part of this PR, or we'll do it on merge.

**Review by Hessesian — APPROVED** [review]
All review comments addressed — looks good to merge

**What changed since initial review:**

- **Dead-code in `cst_cursor.rs`** — fixed: redundant early return gone, only multiline byte-offset version remains
- **`doc.rs` conflict** — dropped cleanly; already in main, no conflict
- **Non-ASCII in `dot_receiver`** — fixed: `c >= 0x80` check alongside `is_ascii_alphanumeric()` handles Unicode identifiers
- **Sort tier (`a:`)** — `collect_stdlib` unconditionally overwrites `sort_text` to `3X:` tier on every item, so `a:` in `build_bare_completions` never wins. Concern was moot — keywords correctly land in tier 3.
- **`this`/`super`** — correctly dropped; only `true`/`false`/`null` in final list
- **Ignored regression tests** — `#[ignore]` removed from both `regression_126_*` tests; they now pass

The `resolve_dotted_receiver_type` chain resolution and multiline call-site fallback are the core wins. Clean implementation.

**Comment by Hessesian** [comment]
Just want to add, thanks a lot for contributing, issues and PR's are very welcome!

**Comment by Hessesian** [comment]
CI is failing on two clippy errors — quick fixes needed before merge:

**1. `src/features/completion.rs:24` — unused import**
```
error: unused import: `crate::StrExt`
```
The `dot_receiver` refactor no longer calls `starts_with_uppercase()` / `starts_with_lowercase()` through `StrExt`. Remove the import:
```rust
// delete this line:
use crate::StrExt;
```

**2. `src/indexer/infer/cst_cursor.rs:115` — needless borrow**
```
error: this expression creates a reference which is immediately dereferenced by the compiler
  --> cst_cursor.rs:115:37
   |
   |         return text_based_call_info(&full_text, cursor_byte);
   |                                     ^^^^^^^^^^^ help: change this to: `full_text`
```
`text_based_call_info` takes `&str`, and `full_text: String` coerces automatically — drop the `&`:
```rust
return text_based_call_info(full_text, cursor_byte);
```

Both are one-line fixes.

**Comment by ulite-Amr** [comment]
Hi @Hessesian,

Thank you so much for the detailed and constructive code reviews! 

This is actually my very first Pull Request in open-source. I learned a lot from your feedback—especially about maintaining clean naming standards, avoiding variable shortcuts, and handling edge cases like non-ASCII identifiers and nullability safely.

I have addressed both Clippy issues (the unused import and the needless borrow) and verified locally that `cargo test` and `cargo clippy` are now fully passing green. 

I'm also incredibly excited about the roadmap you mentioned, especially the upcoming plans for auto-indexing Gradle-cached libraries and handling Kotlin metadata. It will be a massive game-changer for non-IntelliJ editors.

Thank you again for your patience, guidance, and for making my first open-source contribution such a rewarding experience!
